File: soc/jh7110/src/drivers/sbi.c
/*
 * drivers/sbi.c — SBI (Supervisor Binary Interface) Driver for JH7110
 * =====================================================================
 *
 * THIS IS NEARLY IDENTICAL TO soc/kyx1/drivers/sbi.c.
 *
 * Both the Ky X1 and JH7110 run OpenSBI in M-mode. Our kernel runs in
 * S-mode. The SBI specification is architecture-defined — the ecall
 * convention, extension IDs, and function IDs are the same regardless
 * of the underlying SoC.
 *
 * The only reason this is a separate file (rather than a shared sbi.c
 * in common/) is that it uses JH7110-specific constants for CPU vendor
 * and arch ID validation, and we keep each SoC directory self-contained.
 *
 * SBI OVERVIEW:
 * =============
 * SBI (Supervisor Binary Interface) is the RISC-V equivalent of BIOS/UEFI
 * for M-mode to S-mode communication. OpenSBI implements it.
 *
 * We call SBI via ecall with:
 *   a7 = Extension ID (EID)
 *   a6 = Function ID (FID) — 0 for legacy extensions
 *   a0-a5 = Arguments
 *   Return: a0 = error code, a1 = value
 *
 * EXTENSIONS WE USE:
 *   0x01        — Legacy console putchar (always available, single char)
 *   0x10 FID0   — SBI specification version probe
 *   0x10 FID1   — Extension probe (check if an extension is present)
 *   0x54494D45  — SBI Timer (set_timer)
 *   0x48534D   — Hart State Management (hart_start, hart_stop)
 *
 * CPU CSR READS:
 *   RISC-V CSRs mvendorid, marchid, mimpid are M-mode only.
 *   We must use the SBI GetMachineVendorID extension (if available)
 *   or read them via the SBI v0.1 CSR extension.
 *
 *   Alternative: OpenSBI exposes these via the RFENCE extension on
 *   newer platforms. The simplest approach for Tutorial-OS is to
 *   read them directly in M-mode during OpenSBI's platform init —
 *   but since we don't modify OpenSBI, we use the base SBI calls.
 *
 *   On the JH7110:
 *     mvendorid = 0x489 (SiFive, JEDEC bank 10 ID 0x09)
 *     marchid   = 0x8000000000000007 (U74-MC identifier)
 *
 * SIFIVE U74 vs SPACEMIT X60 IDENTIFICATION:
 *   Both are S-mode kernels calling into OpenSBI. The SBI call sequence
 *   is identical. The CPU identification constants differ:
 *     U74:  mvendorid=0x489, marchid=0x8000000000000007
 *     X60:  mvendorid=0x710, marchid=X60-specific
 *   This is how hal_platform_get_info() can report the right CPU details.
 */

#include "sbi.h"
#include "jh7110_regs.h"
#include "types.h"

/* =============================================================================
 * RAW SBI ECALL
 * =============================================================================
 */

typedef struct {
    long error;
    long value;
} sbi_ret_t;

static inline sbi_ret_t sbi_ecall(long eid, long fid,
                                   long a0, long a1, long a2,
                                   long a3, long a4, long a5)
{
    sbi_ret_t ret;
    register long ra0 __asm__("a0") = a0;
    register long ra1 __asm__("a1") = a1;
    register long ra2 __asm__("a2") = a2;
    register long ra3 __asm__("a3") = a3;
    register long ra4 __asm__("a4") = a4;
    register long ra5 __asm__("a5") = a5;
    register long ra6 __asm__("a6") = fid;
    register long ra7 __asm__("a7") = eid;

    __asm__ volatile(
        "ecall"
        : "+r"(ra0), "+r"(ra1)
        : "r"(ra2), "r"(ra3), "r"(ra4), "r"(ra5), "r"(ra6), "r"(ra7)
        : "memory"
    );

    ret.error = ra0;
    ret.value = ra1;
    return ret;
}

/* =============================================================================
 * SBI BASE EXTENSION (EID 0x10)
 * =============================================================================
 */

long sbi_get_spec_version(void)
{
    sbi_ret_t ret = sbi_ecall(0x10, 0, 0, 0, 0, 0, 0, 0);
    return ret.error == 0 ? ret.value : -1;
}

long sbi_probe_extension(long ext_id)
{
    sbi_ret_t ret = sbi_ecall(0x10, 3, ext_id, 0, 0, 0, 0, 0);
    return ret.error == 0 ? ret.value : 0;
}

long sbi_get_mvendorid(void)
{
    sbi_ret_t ret = sbi_ecall(0x10, 4, 0, 0, 0, 0, 0, 0);
    return ret.error == 0 ? ret.value : -1;
}

long sbi_get_marchid(void)
{
    sbi_ret_t ret = sbi_ecall(0x10, 5, 0, 0, 0, 0, 0, 0);
    return ret.error == 0 ? ret.value : -1;
}

long sbi_get_mimpid(void)
{
    sbi_ret_t ret = sbi_ecall(0x10, 6, 0, 0, 0, 0, 0, 0);
    return ret.error == 0 ? ret.value : -1;
}

/* =============================================================================
 * CPU INFO POPULATION
 * =============================================================================
 */

void jh7110_sbi_get_cpu_info(jh7110_cpu_info_t *info)
{
    if (!info) return;

    info->mvendorid = (uint64_t)sbi_get_mvendorid();
    info->marchid   = (uint64_t)sbi_get_marchid();
    info->mimpid    = (uint64_t)sbi_get_mimpid();

    /* Identify the core: SiFive U74 = mvendorid 0x489, marchid 0x8000000000000007 */
    if (info->mvendorid == JH7110_MVENDORID &&
        info->marchid   == JH7110_MARCHID_U74) {
        info->core_name = "SiFive U74";
        info->isa       = "RV64GC";
    } else {
        info->core_name = "RISC-V (unknown)";
        info->isa       = "RV64GC";
    }

    /* SBI spec version for display */
    long spec = sbi_get_spec_version();
    info->sbi_spec_major = (spec >> 24) & 0x7F;
    info->sbi_spec_minor = spec & 0xFFFFFF;
}

/* =============================================================================
 * SYSTEM CONTROL
 * =============================================================================
 */

void sbi_shutdown(void)
{
    /* SBI System Reset extension (EID 0x53525354) */
    sbi_ecall(0x53525354, 0, 0, 0, 0, 0, 0, 0);
    /* If the above fails, try legacy SBI shutdown */
    sbi_ecall(0x08, 0, 0, 0, 0, 0, 0, 0);
    /* Should not reach here */
    while (1) __asm__ volatile("wfi");
}

void sbi_reboot(void)
{
    /* SBI System Reset: warm reset */
    sbi_ecall(0x53525354, 0, 1, 0, 0, 0, 0, 0);
    while (1) __asm__ volatile("wfi");
}

File: soc/jh7110/src/drivers/sbi.rs
//! SBI (Supervisor Binary Interface) ecall interface.
//!
//! Port of `drivers/sbi.c`.  Provides access to M-mode CSRs (mvendorid,
//! marchid, mimpid) via the SBI Base Extension, and system control via
//! the SBI System Reset Extension.
//!
//! # SBI Extensions Used
//!
//! | EID | Name | Purpose |
//! |-----|------|---------|
//! | 0x10 | Base | Spec version, extension probe, mvendorid/marchid/mimpid |
//! | 0x01 | Legacy Console Putchar | Debug output (see `uart.rs`) |
//! | 0x53525354 | System Reset | Reboot / shutdown |
//!
//! # SiFive U74 vs SpacemiT X60
//!
//! Both are S-mode kernels calling into OpenSBI.  The SBI call sequence
//! is identical.  CPU identification constants differ:
//!   - U74: mvendorid=0x489, marchid=0x8000000000000007
//!   - X60: mvendorid=0x710, marchid=X60-specific

use crate::regs;

/// CPU identification info obtained via SBI ecalls.
///
/// Matches the C `jh7110_cpu_info_t` struct in `sbi.h`.
#[derive(Debug, Clone)]
pub struct CpuInfo {
    pub mvendorid: u64,
    pub marchid: u64,
    pub mimpid: u64,
    pub core_name: &'static str,
    pub isa: &'static str,
    pub sbi_spec_major: u32,
    pub sbi_spec_minor: u32,
}

impl Default for CpuInfo {
    fn default() -> Self {
        Self {
            mvendorid: 0,
            marchid: 0,
            mimpid: 0,
            core_name: "Unknown",
            isa: "RV64GC",
            sbi_spec_major: 0,
            sbi_spec_minor: 0,
        }
    }
}

// ============================================================================
// Raw SBI Ecall
// ============================================================================

struct SbiRet {
    error: i64,
    value: i64,
}

/// Issue an SBI ecall with extension ID, function ID, and up to 6 arguments.
///
/// Matches the C `sbi_ecall(eid, fid, a0, a1, a2, a3, a4, a5)` exactly.
#[inline]
fn sbi_ecall(eid: i64, fid: i64, a0: i64, a1: i64, a2: i64,
             a3: i64, a4: i64, a5: i64) -> SbiRet {
    let error: i64;
    let value: i64;
    unsafe {
        core::arch::asm!(
            "ecall",
            inout("a0") a0 => error,
            inout("a1") a1 => value,
            in("a2") a2,
            in("a3") a3,
            in("a4") a4,
            in("a5") a5,
            in("a6") fid,
            in("a7") eid,
            options(nomem, nostack),
        );
    }
    SbiRet { error, value }
}

// ============================================================================
// SBI Base Extension (EID 0x10)
// ============================================================================

const SBI_EXT_BASE: i64 = 0x10;

/// Get SBI specification version.
///
/// Returns packed value: bits [30:24] = major, bits [23:0] = minor.
pub fn get_spec_version() -> i64 {
    let ret = sbi_ecall(SBI_EXT_BASE, 0, 0, 0, 0, 0, 0, 0);
    if ret.error == 0 { ret.value } else { -1 }
}

/// Probe whether an SBI extension is available.
///
/// Returns nonzero if present, 0 if not.
pub fn probe_extension(ext_id: i64) -> i64 {
    let ret = sbi_ecall(SBI_EXT_BASE, 3, ext_id, 0, 0, 0, 0, 0);
    if ret.error == 0 { ret.value } else { 0 }
}

/// Read mvendorid via SBI Base Extension (FID 4).
pub fn get_mvendorid() -> i64 {
    let ret = sbi_ecall(SBI_EXT_BASE, 4, 0, 0, 0, 0, 0, 0);
    if ret.error == 0 { ret.value } else { -1 }
}

/// Read marchid via SBI Base Extension (FID 5).
pub fn get_marchid() -> i64 {
    let ret = sbi_ecall(SBI_EXT_BASE, 5, 0, 0, 0, 0, 0, 0);
    if ret.error == 0 { ret.value } else { -1 }
}

/// Read mimpid via SBI Base Extension (FID 6).
pub fn get_mimpid() -> i64 {
    let ret = sbi_ecall(SBI_EXT_BASE, 6, 0, 0, 0, 0, 0, 0);
    if ret.error == 0 { ret.value } else { -1 }
}

// ============================================================================
// CPU Info Population
// ============================================================================

/// Populate a [`CpuInfo`] by querying all M-mode CSRs via SBI.
///
/// Matches the C `jh7110_sbi_get_cpu_info()` function.
pub fn get_cpu_info() -> CpuInfo {
    let mvendorid = get_mvendorid() as u64;
    let marchid = get_marchid() as u64;
    let mimpid = get_mimpid() as u64;

    // Identify the core: SiFive U74 = mvendorid 0x489, marchid 0x8000000000000007
    let (core_name, isa) = if mvendorid == regs::MVENDORID
        && marchid == regs::MARCHID_U74
    {
        ("SiFive U74", "RV64GC")
    } else {
        ("RISC-V (unknown)", "RV64GC")
    };

    // SBI spec version
    let spec = get_spec_version();
    let (sbi_spec_major, sbi_spec_minor) = if spec >= 0 {
        (((spec >> 24) & 0x7F) as u32, (spec & 0xFF_FFFF) as u32)
    } else {
        (0, 0)
    };

    CpuInfo {
        mvendorid,
        marchid,
        mimpid,
        core_name,
        isa,
        sbi_spec_major,
        sbi_spec_minor,
    }
}

// ============================================================================
// SBI System Reset Extension (EID 0x53525354 "SRST")
// ============================================================================

const SBI_EXT_SRST: i64 = 0x53525354;
const SBI_LEGACY_SHUTDOWN: i64 = 0x08;

/// Shut down the system.  Falls back to legacy SBI shutdown if SRST unavailable.
pub fn shutdown() -> ! {
    sbi_ecall(SBI_EXT_SRST, 0, 0, 0, 0, 0, 0, 0);
    sbi_ecall(SBI_LEGACY_SHUTDOWN, 0, 0, 0, 0, 0, 0, 0);
    loop { unsafe { core::arch::asm!("wfi", options(nomem, nostack)); } }
}

/// Reboot the system (warm reset).
pub fn reboot() -> ! {
    sbi_ecall(SBI_EXT_SRST, 0, 1, 0, 0, 0, 0, 0);
    loop { unsafe { core::arch::asm!("wfi", options(nomem, nostack)); } }
}
